### PR TITLE
[WIP] Rebindable keybindings

### PR DESCRIPTION
**Pull request by ThatLing**
## Rebindable keybindings
Allows you to rebind keys to a specific action. I wanted to add AZERTY support like we have on yogs but keybindings are hardcoded on the server so I have to do this instead. Currently you can only change your movement keys but I'm planning on adding more actions (macros) in the future. Please suggest any ideas, suggestions or point out any bad code (Which there no doubt is) in the meantime.

image

### Notes
- Arrows keys are always bound to movement and can't be unbound
- A key can only be bound to a single action
- Changing keys is possible in-game
- Disabling hotkeys mode will hide the movement keys, in the future this will only leave the non-hotkey actions

### CODE REVIEW

**Comment by Amelia0010** [comment]
>Please suggest any ideas

Not personally an issue, but a lot of people seem to have an issue with the O key for OOC 😄

**Comment by ThatLing** [comment]
OOC is a macro so that's covered

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -19,6 +19,7 @@
 /code/controllers/failsafe.dm @ThatLing
 /code/controllers/subsystem.dm @ThatLing
 /code/controllers/subsystem/ @ThatLing
+/yogstation/code/controllers/subsystem/ @ThatLing
```
Comment (simonset55) [inline .github/CODEOWNERS:22]: You don't own me foo

File: code/__DEFINES/~yogs_defines/keybindings.dm [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"R",
+	"S",
+	"T",
+	"U",
+	"V",
+	"W",
+	"X",
+	"Y",
+	"Z",
+	"Unbound",
```
Comment (simonset55) [inline code/__DEFINES/~yogs_defines/keybindings.dm:28]: Where is Æ, Ø and Å? WHAT ABOUT THE öäü. Be more tolerant you belgian fuck

File: code/__DEFINES/~yogs_defines/keybindings.dm [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// unbound is broken fix me (:
+
+#define BUTTON_KEY(name, id, dir) \
+	button = movement_keys_inv[num2text(dir)]; \
+	button_bound = TRUE; \
+	if(!button || button == "") \
+		button_bound = FALSE; \
+	dat += "<b>[name]:</b> <a href='?_src_=prefs;preference=keybinding_[id];task=input' [button_bound ? "" : "style='color:red'"]>[button_bound ? button : "Unbound"]</a><br>";
+
+// "uh this looks retarded" - This is calculated when its changed to minimize lag caused by the input SS
```
Comment (simonset55) [inline code/__DEFINES/~yogs_defines/keybindings.dm:39]: No u

File: code/modules/client/preferences.dm [hunk 4]
```diff
@@ -500,7 +501,21 @@ GLOBAL_LIST_EMPTY(preferences_datums)
...
+		// yogs start - Custom keybindings
+		if (2) // Keybindings
+			dat += "<center><a href='?_src_=prefs;preference=hotkeys'>[(hotkeys) ? "Hotkeys" : "Default"]</a></center>"
 
+			dat += "<table><tr><td width='340px' height='300px' valign='top'>"
+			if(hotkeys)
+				var/button
+				var/button_bound
+				BUTTON_KEY("North (Up)", "north", NORTH)
+				BUTTON_KEY("West (Left)", "west", WEST)
```
Comment (simonset55) [inline code/modules/client/preferences.dm:513]: Why do you name two hotkeys after Kanye West's child North West?????

File: code/modules/client/preferences.dm [hunk 5]
```diff
@@ -1313,7 +1333,7 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 				if("hotkeys")
 					hotkeys = !hotkeys
 					if(hotkeys)
-						winset(user, null, "input.focus=true input.background-color=[COLOR_INPUT_ENABLED] mainwindow.macro=default")
+						winset(user, null, "input.focus=true input.background-color=[COLOR_INPUT_DISABLED] mainwindow.macro=default") // yogs - Rebindable keys
```
Comment (simonset55) [inline code/modules/client/preferences.dm:1336]: No turn it back on

File: code/modules/keybindings/bindings_client.dm [hunk 6]
```diff
@@ -5,7 +5,7 @@
 	set hidden = TRUE
 
 	keys_held[_key] = world.time
-	var/movement = SSinput.movement_keys[_key]
+	var/movement = prefs.movement_keys[_key] // yogs - Custom keybinds
```
Comment (simonset55) [inline code/modules/keybindings/bindings_client.dm:8]: I thought we were rebasing why is it called yogs keybinds not tg keybinds lol??????!?

File: yogstation.dme [hunk 7]
```diff
@@ -90,6 +90,7 @@
 #include "code\__DEFINES\vv.dm"
 #include "code\__DEFINES\wall_dents.dm"
 #include "code\__DEFINES\wires.dm"
+#include "code\__DEFINES\~yogs_defines\keybindings.dm"
```
Comment (simonset55) [inline yogstation.dme:93]: Nyaaa~yogs_defines

File: yogstation/code/controllers/subsystem/input.dm [hunk 8]
```diff
@@ -0,0 +1,109 @@
+SUBSYSTEM_DEF(input)
+	name = "Input"
+	wait = 1 //SS_TICKER means this runs every tick
+	init_order = INIT_ORDER_INPUT
+	flags = SS_TICKER
```
Comment (simonset55) [inline yogstation/code/controllers/subsystem/input.dm:5]: Is this the ticker that hides the fact that all admins are part of the schutzstaffel. It places a false flag so the players think we're libcucks?

File: yogstation/code/controllers/subsystem/input.dm [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	priority = FIRE_PRIORITY_INPUT
+	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
```
Comment (simonset55) [inline yogstation/code/controllers/subsystem/input.dm:7]: I don't get this, what are the levels running from and why?

File: yogstation/code/controllers/subsystem/input.dm [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		var/key = oldmode_ctrl_override_keys[i]
+		var/override = oldmode_ctrl_override_keys[key]
+		old_default["Ctrl+[key]"] = "\"KeyDown [override]\""
+		old_default["Ctrl+[key]+UP"] = "\"KeyUp [override]\""
+
+	macro_sets = default_macro_sets
+
+/datum/controller/subsystem/input/proc/setup_default_movement_keys()
+	var/static/list/default_keys = list(
+		"W" = NORTH, "A" = WEST, "S" = SOUTH, "D" = EAST,				// WASD
```
Comment (simonset55) [inline yogstation/code/controllers/subsystem/input.dm:95]: What if i'm using an AZERTY keyboard to get those extra 20 words a minute in?? YOU'RE TRIGGERING ME HAVE SOME RESPECT FOR MY SNOWFLAEK

File: yogstation/code/modules/client/preferences_savefile.dm [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+	if(!movement_keys)
+		movement_keys = SSinput.movement_default
+
+	refresh_keybindings()
+
+/datum/preferences/proc/refresh_keybindings()
+	movement_keys_inv = list()
+	for(var/K in movement_keys)
+		var/V = movement_keys[K]
+		movement_keys_inv[num2text(V)] = K // nice sequential lists
```
Comment (simonset55) [inline yogstation/code/modules/client/preferences_savefile.dm:13]: Thank

**Review by simonset55 — COMMENTED** [review]
How much wood can wood cucker cuck if a wood could wood cukcer

**Comment by ShadowDeath6** [comment]
will this have any effect on players who want to use the current hotkey mode

**Comment by Amelia0010** [comment]
@ShadowDeath6 It shouldn't, since the default hotkeys would be set to the current set

**Pull request closed without merging**